### PR TITLE
Materialize lazy modules fully on load_state_dict instead of leaving them in a partially initialized state.

### LINKED ISSUE #73009: Improve loading for nn.modules.lazy.LazyModuleMixin
### 🚀 The feature, motivation and pitch

Currently LazyModuleMixin always requires user to call `.forward`, so that `LazyModuleMixin._infer_parameters` will be triggered.
However this is not needed when there are not unininitialized parameters due to loading module's state dict from file.

So it will be more meaningfull to check `LazyModuleMixin.has_uninitialized_params` within `LazyModuleMixin._lazy_load_hook`, and if it's false, then remove `_initialize_hook` and `_load_hook` from module, and mutate its `__class__` attribute like in `LazyModuleMixin._infer_parameters`, without calling `module.initialize_parameters(*input)` as it's not needed anymore.

Something like this:
```python
class LazyModuleMixin:
    cls_to_become = None

    def _materialize(self):
        self._initialize_hook.remove()
        self._load_hook.remove()
        delattr(self, '_initialize_hook')
        delattr(self, '_load_hook')
        if self.cls_to_become is not None:
            self.__class__ = self.cls_to_become

    def _lazy_load_hook(
            self, state_dict, prefix, local_metadata, strict,
            missing_keys, unexpected_keys, error_msgs):
        for name, param in itertools.chain(self._parameters.items(), self._buffers.items()):
            key = prefix + name
            if key in state_dict and param is not None:
                input_param = state_dict[key]
                if is_lazy(param):
                    # The current parameter is not initialized but the one being loaded one is
                    # create a new parameter based on the uninitialized one
                    if not is_lazy(input_param):
                        with torch.no_grad():
                            param.materialize(input_param.shape)
 
        if not self.has_uninitialized_params():
           self._materialize()

    def _infer_parameters(self, module, input):
        module.initialize_parameters(*input)
        if module.has_uninitialized_params():
            raise RuntimeError('module {} has not been fully initialized'.format(self._get_name()))
        self._materialize()
```

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by substobeme**
Materialize lazy modules fully on load_state_dict instead of leaving them in a partially initialized state, covering all 13 built-in lazy module types.(Fixes #73009)
- Factoring finalization into reusable _materialize() on LazyModuleMixin.
- Added _lazy_load_pre_materialize_hook() for subclasses to set metadata (in_features, in_channels, num_features) on the load path.
- Updated tests to assert immediate finalization; added forward-after-load tests.

**Note:** Changes _infer_parameters body, which torchrec tests for consistency.

### CODE REVIEW

**Comment by substobeme** [comment]
@vishalgoyal316 Can you please review my PR. Thank you.

File: torch/nn/modules/lazy.py [hunk 1]
```diff
@@ -226,6 +234,29 @@ def _lazy_load_hook(
                         with torch.no_grad():
                             param.materialize(input_param.shape)
 
+        # _materialize removes hooks, but only for future calls; the current load completes normally.
+        if not self.has_uninitialized_params():
+            self._lazy_load_pre_materialize_hook()
+            self._materialize()
```
Comment (vishalgoyal316) [inline torch/nn/modules/lazy.py:242]: not has_uninitialized_params() is not enough to call _materialize() on load. 

Example: LazyInstanceNorm2d(affine=False, track_running_stats=False) never creates UninitializedParameter/UninitializedBuffer (batchnorm.py 257–267), so has_uninitialized_params() is False at init. 
Any load_state_dict (even {}) would finalize with num_features=0 and remove the forward hook. 

Fix: only _materialize() if this load materialized at least one lazy tensor or metadata is complete after _lazy_load_pre_materialize_hook() (e.g. in_features / in_channels / num_features > 0).

File: torch/nn/modules/batchnorm.py [hunk 2]
```diff
@@ -277,6 +277,13 @@ def reset_parameters(self) -> None:
         if not self.has_uninitialized_params() and self.num_features != 0:
             super().reset_parameters()
 
+    def _lazy_load_pre_materialize_hook(self):
+        if self.num_features == 0:
+            if self.affine and isinstance(self.weight, torch.Tensor):
+                self.num_features = self.weight.shape[0]
+            elif self.track_running_stats and isinstance(self.running_mean, torch.Tensor):
+                self.num_features = self.running_mean.shape[0]
```
Comment (vishalgoyal316) [inline torch/nn/modules/batchnorm.py:285]: For affine=False and track_running_stats=False, eager InstanceNorm checkpoints often have no tensors; num_features is not in state_dict. 
The hook cannot set num_features from weight/running_mean.

File: test/nn/test_lazy_modules.py [hunk 3]
```diff
@@ -677,6 +692,19 @@ def test_lazy_batchnorm2d_state(self):
...
+        module.eval()
+        lazy_module = nn.LazyBatchNorm2d()
+        lazy_module.load_state_dict(module.state_dict())
+        lazy_module.eval()
+        self.assertIsInstance(lazy_module, nn.BatchNorm2d)
+        self.assertNotIsInstance(lazy_module, nn.LazyBatchNorm2d)
+        self.assertEqual(lazy_module.num_features, 16)
+
+        x = torch.randn(2, 16, 4, 4)
+        self.assertTrue(torch.allclose(lazy_module(x), module(x)))
```
Comment (vishalgoyal316) [inline test/nn/test_lazy_modules.py:708]: Add an explicit regression: LazyInstanceNorm2d(affine=False, track_running_stats=False) + load_state_dict(InstanceNorm2d(10, ...).state_dict()) and assert correct num_features / lazy-vs-concrete behavior.

**Review by vishalgoyal316 — CHANGES_REQUESTED** [review]

File: torch/nn/modules/lazy.py (hunk 1 above)
Comment (substobeme) [reply torch/nn/modules/lazy.py:242]: Thank you for the review, so for this I am adding a  had_lazy flag such that _materialize()  only fires if the loop actually materialized at least one lazy tensor. Condition will be  now if had_lazy and not self.has_uninitialized_params().

File: test/nn/test_lazy_modules.py (hunk 3 above)
Comment (substobeme) [reply test/nn/test_lazy_modules.py:708]: Thank you, adding a test for this verifying module stays lazy after load and becomes concrete after forward.

**Review by vishalgoyal316 — APPROVED** [review]
LGTM!
